### PR TITLE
chore(css): css preprocessors support enhancement

### PR DESCRIPTION
**Pull request by jkuri**
This addon allows "global" CSS extension language files in `public/` dir which are then merged to `dist/app.css` and that file can be included in `index.html`. 

This webpack build runs concurrently with the main build compilation.

### CODE REVIEW

File: addon/ng2/models/webpack-build-css.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+export function getWebpackCSSConfig(projectRoot: string, publicDir: string) {
+  return {
+    name: 'css',
+    entry: { app: glob.sync(`${projectRoot}/${publicDir}/**/*.{sass,scss,less,styl}`) },
+    resolve: {
+      root: path.resolve(projectRoot, publicDir)
+    },
+    output: {
+      path: path.resolve(projectRoot, './dist'),
+      filename: '[name].css'
```
Comment (sirajc) [inline addon/ng2/models/webpack-build-css.ts]: no `hash` in output filename, _hash_ will be required in prod build

**Comment by sirajc** [comment]
@jkuri CSS/SASS files are part of source for project and it would be good if we have that under `src` folder, something like `src/assets`

**Comment by Meligy** [comment]
+1 to @sirajc. I think, as per the ember-cli convention, the `public` folder should have no processors other than copying aso-is into the root of the website.

**Comment by clydin** [comment]
Instead of a fixed location, how about a configuration setting for global styles?  Ideally the option would accept an array of files with glob support.  This allows `public`, as well as `src/assets`, or just an individual file.  It could also allow pulling in styles directly from a package as well (bootstrap?).

Where to put the option is the question.  Does it fit in angular-cli.json?  Or would it be better to add a separate config file for project specific build options?

**Comment by jkuri** [comment]
I'll update this in the next hours, but it is not quite sure this will be merged.

**Comment by filipesilva** [comment]
We talked a bit more about this functionality, and this is the current plan:
- add the following property to config
```
{
   // ...
  "apps": [
    {
     // ...
      styleEntryPoints: {
        "src/style.css": { output: "style.css", autoImported: true }
      }
    }
  ],
```
- add file on sourceDir, maybe with a comment about css imports if the styleExt for the project is css
- add entry point on wepack config common using config to resolve
- add test that checks that global styles are included (see test for sass/etc for code you can copy paste)

**Comment by intellix** [comment]
We have component-name.css so why not index.css next to index.html?

File: addon/ng2/models/webpack-build-common.ts [hunk 2]
```diff
@@ -73,7 +73,11 @@ export function getWebpackCommonConfig(projectRoot: string, sourceDir: string) {
         filename: 'inline.js',
         sourceMapFilename: 'inline.map'
       }),
-      new CopyWebpackPlugin([{from: path.resolve(projectRoot, './public'), to: path.resolve(projectRoot, './dist')}])
+      new CopyWebpackPlugin([{
+        from: path.resolve(projectRoot, './public'), 
+        to: path.resolve(projectRoot, './dist'),
+        ignore: ['**/*.sass', '**/*.scss', '**/*.styl', '**/*.less']
+      }])
```
Comment (sirajc) [inline addon/ng2/models/webpack-build-common.ts]: @jkuri public directory should be copied as is, in full. styles should be part of `src`

Comment (jkuri) [inline addon/ng2/models/webpack-build-common.ts]: thanks, reverted.

**Comment by jkuri** [comment]
@intellix you could change that in `angular-cli.json`, I'll write a documentation about it. Btw, just pushed to this.

**Comment by intellix** [comment]
shouldn't the file be: `style.__styleext__` to support pre-processors as well?

**Comment by jkuri** [comment]
yes, thanks, was a long night.

**Comment by sirajc** [comment]
Great work @jkuri, Thanks..., with this we can define styles like
```json
"apps": [
    {
      "main": "src/main.ts",
      "tsconfig": "src/tsconfig.json",
      "mobile": false,
      "styles": [
            "assets/vendor/paper.bootstrap.css",
            "styles/index.scss"
      ]
    }
  ]
```
:+1:

**Comment by jkuri** [comment]
@sirajc no problem. I still have some minor improvements to do with injecting styles and write tests but this, I think, would be it. Actually the config looks like:
```js
"styles": {
  "src/style.css": { "output": "style.css", "autoImported": true },
  "src/app.sass": { "output": "app.css", "autoImported": true }
}
```
but this will be noted in the readme for sure. In the meanwhile feel free to test this and report any bugs. Thanks.

**Comment by jkuri** [comment]
I managed to get mobile tests passing and working with app-shell prerender. 
@filipesilva, @TheLarkInn please review.

File: addon/ng2/models/webpack-build-common.ts [hunk 3]
```diff
@@ -60,10 +60,6 @@ export function getWebpackCommonConfig(projectRoot: string, sourceDir: string) {
     },
     plugins: [
       new ForkCheckerPlugin(),
-      new HtmlWebpackPlugin({
```
Comment (filipesilva) [inline addon/ng2/models/webpack-build-common.ts]: Why is this removed?

File: addon/ng2/blueprints/ng2/files/__path__/index.html [hunk 4]
```diff
@@ -22,9 +21,15 @@
   <link rel="apple-touch-icon" sizes="180x180" href="/icons/apple-touch-icon-180x180.png">
   <link rel="apple-touch-startup-image" href="/icons/apple-touch-icon-180x180.png">
   <% } %>
-  
+  <% if (!arguments[0].files) {
```
Comment (filipesilva) [inline addon/ng2/blueprints/ng2/files/__path__/index.html]: What is the purpose of this templating logic? Webpack entry points together with the `HtmlWebpackPlugin` should take care of inserting the css files here.

File: addon/ng2/models/webpack-build-common.ts (hunk 3 above)
Comment (jkuri) [inline addon/ng2/models/webpack-build-common.ts]: We have a custom plugin for that which injects `.css` files configured with `autoImported: true` into `index.html`. In this plugin we have also configuration for mobile support which generates content automactically (app-shell).

File: addon/ng2/models/webpack-build-mobile.ts [hunk 5]
```diff
@@ -2,7 +2,6 @@ import * as webpack from 'webpack';
 import * as path from 'path';
 import * as OfflinePlugin from 'offline-plugin';
 import * as CopyWebpackPlugin from 'copy-webpack-plugin';
-import { PrerenderWebpackPlugin } from '../utilities/prerender-webpack-plugin.ts';
```
Comment (filipesilva) [inline addon/ng2/models/webpack-build-mobile.ts]: This is core mobile functionality, it shouldn't be removed.

File: addon/ng2/models/webpack-config.ts [hunk 6]
```diff
@@ -40,20 +45,34 @@ export class NgCliWebpackConfig {
...
+      conf.plugins.unshift(assetsPluginInstance);
+      conf.plugins.unshift(htmlCliPlugin);
+    });
   }
 
   generateConfig(): void {
     switch (this.target) {
       case "development":
-        this.config = webpackMerge(this.webpackBaseConfig, this.webpackDevConfigPartial);
+        this.config = [
```
Comment (filipesilva) [inline addon/ng2/models/webpack-config.ts]: I thought we talked about not having multiple configs. What changed?

File: addon/ng2/blueprints/ng2/files/__path__/index.html (hunk 4 above)
Comment (jkuri) [inline addon/ng2/blueprints/ng2/files/__path__/index.html]: I believe with our config and logic of injection styles we cannot use `HtmlWebpackPlugin`. Otherwise this line is used because template is going through the lodash `_.template` twice.. first when generating blueprint and then second when generating `index.html` for the `dist/`.

File: addon/ng2/models/webpack-build-mobile.ts (hunk 5 above)
Comment (jkuri) [inline addon/ng2/models/webpack-build-mobile.ts]: I merged that functionallity in our "hmtl" plugin.

File: addon/ng2/models/webpack-config.ts [hunk 7]
```diff
@@ -40,20 +45,34 @@ export class NgCliWebpackConfig {
...
 
     this.generateConfig();
-    this.config.plugins.unshift(new NgCliEnvironmentPlugin({
+    this.config[0].plugins.unshift(new NgCliEnvironmentPlugin({
       path: path.resolve(this.ngCliProject.root, `./${sourceDir}/app/environments/`),
       src: 'environment.ts',
       dest: `environment.${this.environment}.ts`
     }));
+
+    const assetsPluginInstance = new AssetsPlugin({ filename: 'cli.assets.json' });
```
Comment (filipesilva) [inline addon/ng2/models/webpack-config.ts]: What is the purpose of this `cli.assets.json` file?

Comment (jkuri) [inline addon/ng2/models/webpack-config.ts]: It temporarily stores all the assets in a file so I know which files needs to be included.
Please see https://webpack.github.io/docs/long-term-caching.html

**Comment by jkuri** [comment]
this PR does not follow the proper design so I am closing it.

**Comment by sirajc** [comment]
@jkuri Should we discuss the design on #1459
